### PR TITLE
Added example code for 'DynamicCommandDescriptionProvider' and 'DynamicCommandDescriptionProvider'

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added example code for `DynamicCommandDescriptionProvider` and `DynamicCommandDescriptionProvider`

Note: The `BaseDynamicCommandDescriptionProvider` is not yet ready. It is WIP.

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
Do you think this part is sufficient? Or do we have to add more details (e.g. how to reference in the `ThingHandlerFactory` and pass it to the `ThingHandler`)?

**Comment by davidgraeff** [comment]
Yes a ThingHandler example referencing those services is what I like to have here.

How much of a WIP status has `BaseDynamicCommandDescriptionProvider`. Like in soonish to be merged? And why is it "BaseDynamic" and not just "Base". A command description provider is always "dynamic" isn't it.

**Comment by cweitkamp** [comment]
I will add it.

The code for the `BaseDynamicCommandDescriptionProvider` is ready. Did not add a unit test yet. I think I will submit a PR in the evening.

I agree, "Dynamic" is not necessary. But does it really hurt?

**Comment by davidgraeff** [comment]
Too much enterprise feeling for normal binding developers. But who doesn't need an `ExtendedDynamicManagementStrategyCanditateProvider` of some sort :D (https://projects.haykranen.nl/java/)

**Comment by cweitkamp** [comment]
The `BaseDynamicCommandDescriptionProvider` feature has been merged yesterday (see https://github.com/openhab/openhab-core/pull/771).

I added the requested code snippet.

> https://projects.haykranen.nl/java/

:rofl: I will try to keep it shorter in the future

File: developers/bindings/thing-xml.md [hunk 1]
```diff
@@ -389,6 +389,33 @@ public class ExampleDynamicCommandDescriptionProvider implements DynamicCommandD
 }
 ```
 
+Most of the times handlers need to modify those dynamic information.
+Therefore the `ThingHandlerFactory` has to reference the the bundle instance and pass it to the handler.
+
+```java
+public class ExampleHandlerFactory extends BaseThingHandlerFactory {
+
+    private ExampleDynamicStateDescriptionProvider stateDescriptionProvider;
```
Comment (davidgraeff) [inline developers/bindings/thing-xml.md]: You can use OSGi DS 1.3 field references:
```java
@Reference
private ExampleDynamicStateDescriptionProvider stateDescriptionProvider;
```

File: developers/bindings/thing-xml.md [hunk 2]
```diff
@@ -389,6 +389,33 @@ public class ExampleDynamicCommandDescriptionProvider implements DynamicCommandD
 }
 ```
 
+Most of the times handlers need to modify those dynamic information.
+Therefore the `ThingHandlerFactory` has to reference the the bundle instance and pass it to the handler.
```
Comment (davidgraeff) [inline developers/bindings/thing-xml.md]: ...the the...

**Review by davidgraeff — APPROVED** [review]

**Pull request merged**